src/compile.rs

```diff
@@ -459,6 +459,197 @@ pub fn validate_rules(statements: &BTreeSet<ValidatedStatement>) -> Result<(), C
+            if let Statement::Call(call) = call {
+                if let Some(inner_func) = tmp_functions.get(&call.get_cil_name()) {
+                    if !inner_func.is_castable {
+                        num_changed += 1;
+                        func.is_castable = false;
+                        continue 'outer;
+                    }
+                // The call may be a built-in so check the args
+                } else {
+                    for arg in &call.args {
+                        if let Argument::Var(arg) = &arg.0 {
+                            // Need to special case this.*
+                            if arg.to_string().contains("this.") {
+                                num_changed += 1;
+                                func.is_castable = false;
+                                continue 'outer;
+                            }
+                            if let Some(ti) = types.get(arg.as_ref()) {
+                                if ti.is_associated_resource(types) {
+                                    num_changed += 1;
+                                    func.is_castable = false;
+                                    continue 'outer;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    num_changed
+}
+
+pub fn initialize_terminated<'a>(
+    functions: &'a FunctionMap<'a>,
+) -> (Vec<String>, Vec<FunctionInfo<'a>>) {
+    let mut term_ret_vec: Vec<String> = Vec::new();
+    let mut nonterm_ret_vec: Vec<FunctionInfo> = Vec::new();
+
+    for func in functions.values() {
+        let mut is_term = true;
+
+        for call in func.original_body {
+            if let Statement::Call(call) = call {
+                match call.check_builtin() {
+                    Some(_) => {
+                        continue;
+                    }
+                    None => {
+                        is_term = false;
+                        break;
+                    }
+                }
+            }
+        }
+        if is_term {
+            term_ret_vec.push(func.get_cil_name().clone());
+        } else {
+            nonterm_ret_vec.push(func.clone());
+        }
+    }
+
+    (term_ret_vec, nonterm_ret_vec)
+}
+
+pub fn search_for_recursion(
+    terminated_list: &mut Vec<String>,
+    functions: &mut Vec<FunctionInfo>,
+) -> Result<(), CascadeErrors> {
+    let mut removed: u64 = 1;
+    while removed > 0 {
+        removed = 0;
+        for func in functions.clone().iter_mut() {
+            let mut is_term = true;
+            for call in func.original_body {
+                if let Statement::Call(call) = call {
+                    let mut call_cil_name = call.get_cil_name();
+                    // If we are calling something with this it must be in the same class
+                    // so hand place the class name
+                    if call_cil_name.contains("this-") {
```

Comment:
Rather than doing all the string manipulation manually, I'd just copy this bit from `ValidatedCall::new()`:

```
let tmp_str = match types
                    .get(convert_class_name_if_this(class_name, parent_type)?.as_ref())
                {
                    Some(type_name) => get_cil_name(Some(&type_name.name), &call.name),
                    None => call.get_cil_name(), // Expected to error out below
                };
```